verified-contracts/src/main/scala/org/ergoplatform/contracts/Edex.scala

```diff
@@ -0,0 +1,90 @@
+package org.ergoplatform.contracts
+
+import org.ergoplatform.sigma.verified._
+import stainless.annotation.ignore
+
+import scala.language.{implicitConversions, postfixOps}
+
+@ignore
+sealed abstract class Edex extends SigmaContract {
+
+  def buyOrder(
+    ctx: Context,
+    tokenId: Coll[Byte],
+    tokenAmount: Long,
+    buyerPk: SigmaProp
+  ): SigmaProp = {
+    import ctx._
+    buyerPk || {
+      (
+        OUTPUTS.nonEmpty &&
+        OUTPUTS(0).R4[Coll[Byte]].isDefined &&
+        OUTPUTS(0).tokens.nonEmpty
+      ) && {
+        val boxWithTokens     = OUTPUTS(0)
+        val tokens            = boxWithTokens.tokens(0)
+        val tokenIdIsCorrect  = tokens._1 == tokenId
+        val givesEnoughTokens = tokens._2 >= tokenAmount
+        val choosesMe         = boxWithTokens.R4[Coll[Byte]].get == SELF.id
+        val iCanSpend         = boxWithTokens.propositionBytes == buyerPk.propBytes
+
+        choosesMe && tokenIdIsCorrect && givesEnoughTokens && iCanSpend
+      }
+    }
+  }
+
+  def sellOrder(
+    ctx: Context,
+    tokenId: Coll[Byte],
+    tokenPrice: Long,
+    sellerPk: SigmaProp
+  ): SigmaProp = {
+    import ctx._
+    sellerPk || {
+      (
+        OUTPUTS.size > 1 &&
+        OUTPUTS(1).R4[Coll[Byte]].isDefined &&
+        SELF.tokens.nonEmpty
+      ) && {
+        val boxWithCoins     = OUTPUTS(1)
+        val tokenForSale     = SELF.tokens(0)
+        val tokenIdIsCorrect = tokenForSale._1 == tokenId
+        val choosesMe        = boxWithCoins.R4[Coll[Byte]].get == SELF.id
+
+        val givesEnoughCoins = boxWithCoins.value >= (tokenForSale._2 * tokenPrice)
```

Comment:
Rather than specifying an asking price for **all** token amount I use `tokenPrice` (for one token). This should make the contract usable with any token amount, thus "postponing" the decision on how many tokens to sell until building the box.